### PR TITLE
Feat: Add customer search controller

### PR DESCRIPTION
**Pull request by pKorsholm**
**What**
- Add a `controllers` folder to act as a middlelayer between the api- and service layers

**Why**
Adding a middlelayer between the API- and Service layers allows to us abstract api logic away from the specific file and encapsulate endpoint behaviour in a controller file which allows us to repurpose implementations and minimise code duplication in the codebase. 

Initially we will probably structure the controllers as in this pr, but as the controllers for each domain grows we are able to create a sub directory where each function can have its own file and exports without changing the imports of the controllers across the application. 

*Why not just use the service layer*
The controllers layer will act as a separate layer neither api or service layer, but will be able to modifying the received request and build the correct inputs for the service layer as seen in `controllers/customers.ts`. 
We will probably not need a controller method for all the individual api endpoints since some, like delete and create can be very small, but in some cases, like listing, having the code separated will allow us to reuse the code for different endpoints, i.e. list products which exist in both `/store` and `/admin`, where the primary difference is that `/store` only returns published products. By having a controller layer the `/store` endpoint would hardcode the status value of the query before proceeding but then use the same handler as the `/admin` endpoint.

If you have any suggestions please pitch in

### CODE REVIEW

**Comment by adrien2p** [comment]
Iam not sure its a good approach, this approach to share the code for different endpoint break the responsibility principle. I mean by that than the controller method can be updated by two different needs. What happen if there is a new use case fro the store route to take a new params that needs to update the common method, then we will have to check that it is either requested by the store or the admin and therefore mix two different needs that looks the same at first glance but in the end are just different.

**Comment by adrien2p** [comment]
Generally speaking, you would have either an utility method, or a methodA and methodB that both call methodC but A and B manage there logic to call C and not the opposite where methodC manage the logic for both A and B. I dont know if I am clear 😅

**Comment by pKorsholm** [comment]
Hi Adrien

You are very right in that if a new use-case is needed for the store route, and that is not expressed in the general list parameters, then we shouldn't add a switch parameter or any kind of specialised case handling or branching. The correct way to handle that case would be to create a new controller method for the specialised case which would be solely responsible for that case. 

I think that this change actually reflect the method A, B and C which you describe. In which A and B are the individual endpoints, to stick with the example list product in store and admin, and C would be the controller responsible for listing the individual products. A and B both handle their internal logic as they wish, adding a status filter for the store, but then both invoke C which just fetches the product.
I would argue that this doesn't break the single responsibility principle, since the code now in `controllers/customer.ts` is only responsible for listing customers given the list parameters as input. What those parameters are, is determined by the context in which the method in invoked. i.e. in the `list-products` example the input parameters would always be filterable parameters along with generalised pagination. The difference between the two endpoints would be that the `/store` endpoint always limits the products to published products. In that case the work that is done in the controller method would always be performed in the same way, not branching out depending on each use case.



To give a bit of background this change originated from the idea of introducing customer groups. However we wanted a specific endpoint like `/admin/customer-groups/{id}/customers`, which would allow us to fetch customers from a specific group with pagination because a customer group potentially could have many thousands of customers, and fetching them all at once would be a very expensive operation. However this endpoint would have a similar relationship to `list-customers` as `list-products` has in the `/store` to `/admin` endpoints. It would be a `list-customers` endpoint with the only constraint being a single group, which was what catalysed a refactor from the api layer containing the logic for to a controller layer.

**Comment by pKorsholm** [comment]
Hi Adrien

You are very right in that if a new use-case is needed for the store route, and that is not expressed in the general list parameters, then we shouldn't add a switch parameter or any kind of specialised case handling or branching. The correct way to handle that case would be to create a new controller method for the specialised case which would be solely responsible for that case. 

I think that this change actually reflect the method A, B and C which you describe. In which A and B are the individual endpoints, and C would be the controller responsible for listing the individual products. A and B both handle their internal logic as they wish, adding a specific filter for their use case, fetch a product etc., but then both invoke C which just fetches the resulting entities based on whatever input it is given, no specialised cases. Specialised cases can be handled, as you state, in the endpoints or A and B. 

In cases which can't be handled by controlling the input to the controller, the controller should not be used, instead a new controller should be made or if the use case affects both endpoints (say we wanted to add sorting 😉 ) the controller could be modified to comply.

I would argue that this doesn't break the single responsibility principle, since the code now in `controllers/customer.ts` is only responsible for listing customers given the list parameters as input. What those parameters are, is determined by the context in which the method in invoked. i.e. in the `list-products` example the input parameters would always be filterable parameters along with generalised pagination. The difference between the two endpoints would be that the `/store` endpoint always limits the products to published products. In that case the work that is done in the controller method would always be performed in the same way, not branching out depending on each use case.

**Comment by adrien2p** [comment]
Thank you for your answer, I just wanted to point that to keep it in mind 🥰 because that pattern can easily goes wrong. Also, ive got a feeling than there is one more layer which is not necessary. I mean, that example looks like an utility which could handle the arguments formatting for. but maybe Iam missing some input and should wait to see what's coming next 😅

File: packages/medusa/src/api/routes/admin/customers/list-customers.ts [hunk 1]
```diff
@@ -28,26 +27,9 @@ export default async (req, res) => {
...
-    skip: validated.offset,
-    take: validated.limit,
-  }
-
-  const [customers, count] = await customerService.listAndCount(
-    selector,
-    listConfig
+  const [customers, count] = await customerController.listAndCount(
+    customerService,
+    validated
```
Comment (adrien2p) [inline packages/medusa/src/api/routes/admin/customers/list-customers.ts:32]: Would it be nice to rename that variable as validatedQueryParams in order to be more understandable?

File: packages/medusa/src/controllers/customers.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    take: queryparams.limit,
+  }
+
+  const filterableFields = omit(queryparams, ["limit", "offset", "expand"])
+
+  return await customerService.listAndCount(
+    pickBy(filterableFields, (val) => typeof val !== "undefined"),
+    listConfig
+  )
+}
```
Comment (srindom) [inline packages/medusa/src/controllers/customers.ts:27]: I think we should try to standardize the inputs to "controller" functions; could be something like this:

- `scope`, `queryParams`, `payload`, `context`; so when called from the endpoint handler you'd pass 

```
listCustomers(req.scope, req.query, req.body, { ... we'll figure out what should go in here })
```

- I also think that the return type of this should correspond to the `[domain]Res` type i.e. for this it should be:
```
{
  customers: [...],
  count,
  limit,
  offset
}
```

**Review by srindom — CHANGES_REQUESTED** [review]

**Review by srindom — APPROVED** [review]

**Pull request merged**